presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.predicate.Domain;
+import com.facebook.presto.spi.predicate.TupleDomain;
+import com.facebook.presto.spi.predicate.ValueSet;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeUtils;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterSourceOperator
+        implements Operator
+{
+    public static class DynamicFilterSourceOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final List<Type> types;
+        private final List<Integer> filterChannels;
+        private final List<String> filterDynamicFilterNames;
+        private final Optional<DynamicFilterClientSupplier> dynamicFilterClientSupplier;
```

Comment:
The same problem is with `ExchangeClient`.
`LocalQueryRunner` takes `null` for `exchangeClientSupplier`, the `ExchangeOperatorFactory` has check for nullness of `exchangeClientSupplier`. It's pretty much the same behavior I did here but I used `Optional` instead of `null`, which I believe is nicer because forces the user to think about and handle properly the situation of absence of the supplier.